relay-server/src/actors/global_config.rs

```diff
@@ -0,0 +1,230 @@
+        Self::Subscribe(sender)
+    }
+}
+
+/// Service implementing the [`GlobalConfigManager`] interface.
+///
+/// The service offers two alternatives to fetch the [`GlobalConfig`]:
+/// responding to a [`Get`] message with the config for one-off requests, or
+/// subscribing to updates with [`Subscribe`] to keep up-to-date.
+#[derive(Debug)]
+pub struct GlobalConfigService {
+    config: Arc<Config>,
+    /// Sender of the [`watch`] channel for the subscribers of the service.
+    sender: watch::Sender<Arc<GlobalConfig>>,
+    /// Sender of the internal channel to forward global configs from upstream.
+    internal_tx: UnboundedSender<UpstreamResult>,
+    /// Receiver of the internal channel to forward global configs from upstream.
+    internal_rx: UnboundedReceiver<UpstreamResult>,
+    /// Upstream service to request global configs from.
+    upstream: Addr<UpstreamRelay>,
+    /// Handle to avoid multiple outgoing requests.
+    fetch_handle: SleepHandle,
+}
+
+impl GlobalConfigService {
+    /// Creates a new [`GlobalConfigService`].
+    pub fn new(config: Arc<Config>, upstream: Addr<UpstreamRelay>) -> Self {
+        let (sender, _) = watch::channel(Arc::default());
+        let (internal_tx, internal_rx) = mpsc::unbounded_channel();
+        Self {
+            config,
+            sender,
+            internal_tx,
+            internal_rx,
+            upstream,
+            fetch_handle: SleepHandle::idle(),
+        }
+    }
+
+    /// Handles messages from external services.
+    fn handle_message(&self, message: GlobalConfigManager) {
+        match message {
+            GlobalConfigManager::Get(sender) => {
+                sender.send(self.sender.borrow().clone());
+            }
+            GlobalConfigManager::Subscribe(sender) => {
+                sender.send(self.sender.subscribe());
+            }
+        }
+    }
+
+    /// Schedules the next global config request.
+    fn schedule_fetch(&mut self) {
+        if self.fetch_handle.is_idle() {
+            // XXX(iker): set a backoff mechanism?
+            self.fetch_handle
+                .set(self.config.global_config_fetch_interval());
+        }
+    }
+
+    /// Requests a new global config from upstream.
+    ///
+    /// As the upstream service is responsible for authentication, the global
+    /// config service doesn't deal with it.
+    fn update_global_config(&mut self) {
+        self.fetch_handle.reset();
+
+        if !self.config.has_credentials() {
+            return;
+        }
+
+        let upstream_relay: Addr<UpstreamRelay> = self.upstream.clone();
+        let internal_tx = self.internal_tx.clone();
+
+        tokio::spawn(async move {
+            let query = GetGlobalConfig::query();
+            let res = upstream_relay.send(SendQuery(query)).await;
+            if internal_tx.send(res).is_err() {
+                relay_log::error!("failed to forward internally the upstream response");
```

Comment:
if this error happens, it would stop requesting global configs altogether, however, i think it's impossible for this to happen as according to the channel docs, it should only fail if the receiver is dropped, which it won't.